### ISSUE TITLE
Memory bloat in ActiveJob::LogSubscriber

### ISSUE THREAD

**Issue by NobodysNightmare** [rails/rails#46185]
Enqueueing lots of jobs from a running job causes memory to bloat. This seems to be caused by the `ActiveJob::LogSubscriber`, which keeps a stack of all events (e.g. `enqueue.active_job`) that happen during another event (`perform.active_job`).

### Steps to reproduce

This job is written to show the bloat:

```ruby
class TestJob < ApplicationJob
  queue_as :test
  
  def perform
    10.times do
      SecondTestJob.perform_later

      subscriber = ObjectSpace.each_object(ActiveJob::LogSubscriber).first
      dangling_event = subscriber&.send(:event_stack)&.first
      if dangling_event
        puts "Found a dangling event with #{dangling_event.children.size} children."
      else
        puts 'There are no dangling events.'
      end
    end
  end
end
```
It will print `Found a dangling event with X children.`, where `X` is growing with the number of jobs that have been enqueued. This means each time a new job is enqueued, we will store an additional event and not release it until `TestJob` finished performing.

If the code is not excuted as active job (e.g. `TestJob.new.perform`), the output is `There are no dangling events.`, which means no events are accumulated and memory remains stable.

### Expected behavior

I can enqueue an infinite amount of jobs from a job, without running out of memory.

### Actual behavior

Enqueuing jobs from a job is memory constrained, because events will pile up. This is especially a problem, if enqueueing jobs with large parameters, since the parameters will be referenced in memory as well. Example:

```ruby
class TestLargeObjectJob < ApplicationJob
  queue_as :test
  
  def perform
    10.times do
      # large_object  should be garbage collectible after every iteration, but the dangling event will keep a reference to it, so large_object will never be collected
      large_object = 'A' * 1_000_000
      SecondTestJob.perform_later(large_object)
    end
  end
end
```

### System configuration
**Rails version**: 7.0.4

**Ruby version**: 3.0.0

**Comment by skipkayhil** [comment]
It looks like this was addressed on main in #43390, but not backported to 7.0 because it was a breaking change. I'll leave this issue open in case @jhawthorn thinks this could be backported, but you might just have to wait for 7.1 for a fix.

**Comment by NobodysNightmare** [comment]
Thanks for the cross reference. This showed me that the exact same issue as mine reported before (in #21036). For some reason I did not find that, though I was looking.

Since there is a fix and an existing issue for the exact same problem, I am going to close this issue as well.

Thanks.
